src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -840,93 +867,889 @@ class ApplyHelper
+                }
+            }
+            else
+            {
+                meterReads = true;
+                auto ltxe = ltx.loadWithoutRecord(lk);
+                if (ltxe)
+                {
+                    entrySize =
+                        static_cast<uint32_t>(xdr::xdr_size(ltxe.current()));
+                    entryMap.emplace(lk, ThreadEntry{ltxe.current(), false});
+                }
+            }
+
+            if (!meterReads)
+            {
+                continue;
+            }
+
+            uint32_t keySize = static_cast<uint32_t>(xdr::xdr_size(lk));
+            readEntryCounters.noteDiskReadEntry(isContractCodeEntry(lk),
+                                                keySize, entrySize);
+
+            if (resources.diskReadBytes < readEntryCounters.mLedgerReadByte)
+            {
+                innerResult(res).code(
+                    INVOKE_HOST_FUNCTION_RESOURCE_LIMIT_EXCEEDED);
+
+                diagnosticEvents.pushError(
+                    SCE_BUDGET, SCEC_EXCEEDED_LIMIT,
+                    "operation byte-read mresources exceeds amount specified",
+                    {makeU64SCVal(readEntryCounters.mLedgerReadByte),
+                     makeU64SCVal(resources.diskReadBytes)});
+
+                // Only mark on failure because we'll also count read bytes
+                // during apply to also count anything read from the hot
+                // archive.
+                sorobanMetrics.mHostFnOpReadEntry.Mark(
+                    readEntryCounters.mReadEntry);
+
+                sorobanMetrics.mHostFnOpReadKeyByte.Mark(
+                    readEntryCounters.mReadKeyByte);
+                sorobanMetrics.mHostFnOpReadLedgerByte.Mark(
+                    readEntryCounters.mLedgerReadByte);
+                sorobanMetrics.mHostFnOpReadDataByte.Mark(
+                    readEntryCounters.mReadDataByte);
+                sorobanMetrics.mHostFnOpReadCodeByte.Mark(
+                    readEntryCounters.mReadCodeByte);
+
+                sorobanMetrics.mHostFnOpMaxRwKeyByte.Mark(
+                    readEntryCounters.mMaxReadWriteKeyByte);
+                sorobanMetrics.mHostFnOpMaxRwDataByte.Mark(
+                    readEntryCounters.mMaxReadWriteDataByte);
+                sorobanMetrics.mHostFnOpMaxRwCodeByte.Mark(
+                    readEntryCounters.mMaxReadWriteCodeByte);
+
+                return false;
+            }
+        }
+        return true;
+    };
+    bool success = getEntries(mParentTx.sorobanResources().footprint.readOnly);
+    if (success)
+    {
+        success = getEntries(mParentTx.sorobanResources().footprint.readWrite);
     }
+    return success;
 }
 
-bool
-InvokeHostFunctionOpFrame::doApply(
-    AppConnector& app, AbstractLedgerTxn& ltx, Hash const& sorobanBasePrngSeed,
-    OperationResult& res,
-    std::optional<RefundableFeeTracker>& refundableFeeTracker,
-    OperationMetaBuilder& opMeta) const
+class ParallelApplyHelper
```

Comment:
The V23 code can be removed from ParallelApplyHelper. We should also de-duplicate some of the shared code.